acme.sh

```diff
@@ -5265,10 +5267,12 @@ _installcert() {
       cp "$_real_key" "$_backup_path/key.bak"
     fi
     if [ -f "$_real_key" ]; then
+      chmod 600 "$_real_key"
```

Comment:
don't  chmod if the key file is already existing.

Reply:
One of the goal of this patch is to avoid leaking the private key via wrongly set file mode. Without this command, the new private key saved in the subsequent `cat` command may end up in a file that's world readable (if someone set them that way between two `acme.sh` runs). If somebody is intentionally willing to leak the private key, the "best" solution is to copy it somewhere else, but IMO `acme.sh`'s correct behaviour is to always make sure that a saved private keys don't end up in a world-readable file.

Reply:
>> (if someone set them that way between two acme.sh runs). If somebody is intentionally willing to leak the private key, the "best" solution is to copy it somewhere else,

No, we should chmod only if the file doesn't exist.  If the user changed the chmod, we should respect it.

Reply:
You assume that the user does this intentionally and that the user is an actual person. I'd argue this is easier done by mistake by a human user or done by buggy tools. `acme.sh` is a tool designed to create a secure and unattended way to update and handle private keys, so I think it's rightly expected that is enforces security around the files it manages. Also note that the user is not expected to tamper with files managed by `acme.sh` (and similar automated tools.)

Reply:
If the key file doesn't exist,  we create it and chmod to 600. That's ok.
But if the key file already exists,  we should not chmod to the file.  We MUST respect the user's choice.

Reply:
Okay, I won't argue just closing this and move on to a secure client.

Reply:
Some final thoughs in case somebody is interested in securing this ACME client:

1. If this project keeps insisting on "respecting user's choice" for private key file modes, a subsequent fix for the world-readable issue will NEVER be fixed for existing installations, only for new ones, or for newly created certs. The reason is that at this moment `acme.sh` creates files with the wrong (insecure) file mode, which will be kept as-is, even though it was never a user's choice, but a result to buggy (insecure) `acme.sh` behavior. Also, any time in the future there will be a bug in `acme.sh` that results in a wrong file mode or similar, it will similarly be never fixed by subsequent updates, if this "rule" is followed.

2. Insisting on "respecting user's choice" for a file that is a) not created by the user b) managed automatically by this tool c) not supposed to be touched by the user: makes no sense.

3. Respecting user's choice makes absolute sense for files created, edited and managed by users. `acme.sh`-managed private keys aren't such files.

4. If there are `acme.sh` users who require their private keys to be world readable, or relied on the existing buggy behavior, either those environments should be fixed, or a wrapper script created that recreates the insecure state each time after running `acme.sh`. But, the default should be secure, not insecure as it is now.

Reply:
I agree that the private keys file could be  "600",  but only if the file is created by acme.sh.

After we generate the key file, if the user wants to change the mod other than 600,  he could do "chmod" by hand.

When the cert is renewed in the future,  the user's "chmod" will be kept.


So, there is NO secure issue unless the user change the mod by his hand.  
If the user chmod on the key file, he takes responsibility.

All we do is to  keep the cert working when renewal without any damage to the user's system.

Reply:
private key files are recommended to be 600,  but remember that it's just a normal file, it can be any mod.
What the mod is MUST meet the user's needs.
What meets the needs is the best choice.

Reply:
what acme.sh does is never to make/change choices for the user. Instead, we respect the user's choice MOST.

Reply:
You're repeating this notion of respecting the user's choice, but you're ignoring all the issues I've mentioned above that arises from this, in this particular case, including the fact that this isn't a user created file, not is it a "normal file". Note that e.g. ssh will even fail with an error unless the file is 0600. `acme.sh` is lucky that most web servers aren't such picky about this.

To me it looks like `acme.sh` is purposefully designed to be insecure and for some unknown reason maintainers want to keep it that way.

Reply:
>>  but you're ignoring all the issues I've mentioned above

Not true.  I didn't ignore the issues.  And I said I agree with you that the key file can be set to 600 when we first create the file.
The only argument between us is that:  whether to set the key file to 600 every time we renew the cert.

You insist to set to 600 every time, no matter the file is already existing or not.
My point is not to do so if the file is already existing.

And I don't agree with you:  acme.sh is secure enough.


>> acme.sh is lucky that most web servers aren't such picky about this.

you know why is this?  why most web servers aren't such picky?
There must be a balance between Secure and Convenience.  
100% secure is not always the best choice in all the cases.

Reply:
Okay, so I understand you confirmed that `acme.sh` is purposefully designed to be insecure, and there is no desire to fix this issue for existing installations and existing private keys.